File: src/main/java/io/Cbuild/cbuildException.java
package io.Cbuild;

public class cbuildException extends RuntimeException {

    public enum ErrorType {
        SYNTAX,
        SEMANTIC,
        PROCESS
    }

    private final ErrorType errorType;
    private final String fileName;
    private final int line;
    private final int column;

    public cbuildException(ErrorType errorType, String message) {
        super(message);
        this.errorType = errorType;
        this.fileName = null;
        this.line = -1;
        this.column = -1;
    }

    public cbuildException(ErrorType errorType, String message, String fileName, int line, int column) {
        super(message);
        this.errorType = errorType;
        this.fileName = fileName;
        this.line = line;
        this.column = column;
    }

    public cbuildException(ErrorType errorType, String message, int line, int column) {
        super(message);
        this.errorType = errorType;
        this.fileName = null;
        this.line = line;
        this.column = column;
    }

    public ErrorType getErrorType() {
        return errorType;
    }

    public String getFileName() {
        return fileName;
    }

    public int getLine() {
        return line;
    }

    public int getColumn() {
        return column;
    }

    @Override
    public String getMessage() {
        return getFormattedMessage();
    }

    public String getFormattedMessage() {
        String severity = errorType.name().toLowerCase();
        String rawMessage = super.getMessage();

        if (!hasSourceLocation()) {
            return severity + " error: " + rawMessage;
        }

        return "%s:%d:%d: %s error: %s"
                .formatted(fileName, line, column, severity, rawMessage);
    }

    public String getRawMessage() {
        return super.getMessage();
    }

    private boolean hasSourceLocation() {
        return fileName != null && line >= 0 && column >= 0;
    }
}

File: src/cbuild-exception.ts
export enum ErrorType {
  SYNTAX,
  SEMANTIC,
  PROCESS,
}

export enum MachineCode {
  // compilation
  COMPILATION_ERROR,
  FUNCTION_COMPILATION_ERROR,

  // semantic
  UNSUPPORTED_IR,

  // process
  FILE_NOT_FOUND,
  DEPQ_NOT_FOUND,

  // depq graph
  NO_TARGET_FOUND,
  CIRCULAR_DEPQ,

  // command
  SHELL_COMMAND_FAILED,
  SHELL_COMMAND_ABORTED,
  INVALID_SHELL_PATH,
}

export interface CbuildExceptionOptions {
  message: string;
  errorType: ErrorType;
  machineCode: MachineCode;
  row: number;
  column: number;
}

export class CbuildException extends Error {
  public constructor(
    message: string,
    public readonly errorType: ErrorType,
    public readonly machineCode: MachineCode,
    public readonly row: number,
    public readonly column: number,
  ) {
    super(message);
    this.name = "CbuildException";

    Object.setPrototypeOf(this, new.target.prototype);
  }

  public static from({
    message,
    errorType,
    machineCode,
    row,
    column,
  }: CbuildExceptionOptions): CbuildException {
    return new CbuildException(message, errorType, machineCode, row, column);
  }
}
